### PR TITLE
Dataloaders

### PR DESCRIPTION
**Pull request by hhsecond**
## Motivation and Context
Integration with other popular tools which could be used alongside hangar is really useful for users. Dataloaders is probably one of them.
#13 

## Description
Right now, it's just a skeleton to start the discussion about the structure. I'll change it as it moves forward.

## Types of changes
What types of changes does your code introduce? Put an `x` in all the boxes that apply:
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

Is this PR ready for review, or a work in progress?
- [ ] Ready for review
- [x] Work in progress

## How Has This Been Tested?
Put an `x` in the boxes that apply:
- [ ] Current tests cover modifications made
- [ ] New tests have been added to the test suite
- [ ] Modifications were made to existing tests to support these changes
- [ ] Tests may be needed, but they are not included when the PR was proposed
- [ ] I don't know. Help!

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **[CONTRIBUTING](../CONTRIBUTING.rst)** document.
- [x] I have signed (or will sign when prompted) the tensorwork CLA.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by hhsecond** [comment]
@rlizzo I made the basic skeleton of the dataloader base class. For now, we could keep the other dataloaders also in the same file. Taking your discussion from #13, adding features such as batching, shuffling to the base dataloader is not required as of now since both PyTorch and Tensorflow does that quite well. They just need an object from where they could get one sample at a time. In my view, the base class can be inherited and make advanced dataloaders inside hangar itself when it's required in the future.

**Comment by rlizzo** [comment]
Hey sherin I think we should put this on the back burner right now. The data loaders are important, but the bulk of the work to define multithreaded (multiprocess?) and batch reads actually should be implemented in the hangar core and enabled by default for some operations. These utilities could just implement the framework specific interfaces and make calls straight to the public hangar API. 

I'm also not thrilled about including them in the `hangar-py/src` directory. I would like to keep only User facing APIs and the backend processing workflows in there. Since these sorts of utility functions are just bolted on for convienence (similar to what we want to do with the CLI import/export) I'm thinking they should almost be stored as a separate module? 

What I mean is that we may want to split the `src` dir into `core` (containing what is in the repo right now) and a `convenience` dir (or some other similar name) to hold these. To keep a unified experience for importing the Hangar Repo or Dataloaders classes in a dependent codebases we could just raise the "apparent" level of each module to a unified interface. When used as intended (and not like most home brewed projects do) we can do some really useful things with a modules `__init__.py` file. I can handle all of that when it comes time...

With the changes introduced in #71, multithreaded reading got a whole lot simpler. Let's coordinate on those changes and how the backend now thinks about files before I leave for holiday on Tuesday. In brief though the backend routing is now about as simple as it gets (essentially dynamic backend selection by returning accessor methods as key lookups in a dict, and then blindly passing in a `hash` record value as input to the returned function which then "magically" returns the data.) The implementation details aren't really your concern when operating at the level this would need to work on,  but I'll just say that its rather efficient, and we should have no issues with multiple threads (or even processes) reading at the same time (even if they actually need to read from the same file on disk).

**Comment by hhsecond** [comment]
@rlizzo Awesome, make sense. However, I would like to make a very simple (at least pytorch) data loader for the demo I have on 15th at PyCon Thailand. Then we can probably be iterating over it when we get more time for this. Although a utility in the core would be the best, I can write my own snippet example and tell the audience that something like this is coming into the core soon.

**Comment by rlizzo** [comment]
Yeah if u can just write it as a user defined script on ur local system that would be best for now. Feel free to post it in a gist so we can use that as a jumping off point in the future.

 I don't want to limit you too much or sound too dictatorial, but it's important to me that we keep the code logically ordered (while maintaining ease of use). This is complicated enough as is, I won't accept bloat (I already want to kill some redundancies in the `records/queries.py` and `records/heads.py` and I think some of the functions in `records/parsing.py` are either unreachable or redundant right now...) Hope you can understand. We'll work together to make sure that these Bolton utilities can do what they need while respecting the structure of the core

**Comment by hhsecond** [comment]
Make sense. Closing this for now!
